google_analytics/notebooks/load.ipynb

```diff
@@ -34,24 +34,189 @@
+  {
+   "cell_type": "code",
+   "execution_count": 65,
    "metadata": {},
    "outputs": [],
    "source": [
-    "train = load(\"../data/train.csv\", nrows=100000)\n",
-    "test = load(\"../data/test.csv\", nrows=10000)\n",
+    "x_train = train[(train[\"date\"] < '2016-12-01')]\n",
+    "y_train = train[(train[\"date\"] >= '2016-12-01') & (train[\"date\"] >= '2017-02-01')]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 66,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 263866 such visitors (out of 295432 found in our training set in total).\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Find people with a single visit during the training period\n",
+    "count = x_train[[\"fullVisitorId\", \"transactionRevenue\"]].groupby(\"fullVisitorId\").count().reset_index()\n",
+    "\n",
+    "single_visit_ids = count[count[\"transactionRevenue\"] == 1][\"fullVisitorId\"]\n",
+    "print(\"There are {} such visitors (out of {} found in our training set in total).\".format(len(single_visit_ids), len(count)))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 67,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Out of 263866 single visits, 1107 ended up in a purchase\n"
+     ]
+    }
+   ],
+   "source": [
+    "# But maybe some of those guys actually bought something in that single visit.\n",
+    "single_visits = x_train[x_train[\"fullVisitorId\"].isin(single_visit_ids)]\n",
+    "num_purchases = len(single_visits[single_visits[\"transactionRevenue\"] > 0][\"transactionRevenue\"])\n",
+    "print(\"Out of {} single visits, {} ended up in a purchase\".format(len(single_visit_ids), num_purchases))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "We will therefore choose to ignore 262759 users\n"
+     ]
+    }
+   ],
+   "source": [
+    "to_ignore = set(single_visit_ids) - set(single_visits[single_visits[\"transactionRevenue\"] > 0][\"fullVisitorId\"])\n",
+    "print(\"We will therefore choose to ignore {} users\".format(len(to_ignore)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Is this assumption justified?\n",
+    "\n",
+    "How reliable is this assumption? Are these people that we choose to ignore indeed buying nothing? Let's find out.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 72,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 2290 people appearing both in the training and the test periods\n",
+      "We will ignore 2037 out of those 2290 people\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We will only focus on people who exist in both datasets\n",
```

Comment:
Am I correct that the question we want to answer is: How many people visited once during x_train and did not buy there (people in the list to_ignore) buy during y_train? The lower this number the better we validate our assumption to drop these people?
So I guess the reason why we focus on people existing in both datasets is that people in to_ignore that are not in y_train are definitely not buying?

Reply:
Yes that is indeed the correct question. What I have answered here is slightly different however, because I am only taking into account the intersection of clients (people appearing at least once in both train and test). The denominator in my division should probably be the number of people in train regardless if they appear in test (those who don’t obviously have a 0 revenue). Therefore the actual numbers should be much lower than the ones I report here.

You can fix yourself or I can do it when I get back (Sunday/Monday)

Reply:
No problem! I will change it.